backend/vertical-healthcare_ch/forum_datenaustausch_ch.invoice_450_request/src/main/java/de/metas/vertical/healthcare_ch/forum_datenaustausch_ch/invoice_450/Invoice450FromCrossVersionModelTool.java

```diff
@@ -1505,4 +1515,86 @@ private DocumentType createDocumentType(@NonNull final XmlDocument document)
 
 		return documentType;
 	}
+
+	public XmlRequest augmentRequest(final XmlRequest xAugmentedRequest, final BPartnerId bPartnerId)
+	{
+		if (isValidEsrQRCandidate(xAugmentedRequest))
+		{
+			final String qrIban = getQrIbanOrNull(bPartnerId);
+			if (!Check.isBlank(qrIban))
+			{
+				final XmlBody body = xAugmentedRequest.getPayload()
+						.getBody();
+				return xAugmentedRequest.withMod(RequestMod.builder()
+						.payloadMod(PayloadMod.builder()
+								.bodyMod(BodyMod
+										.builder()
+										.esr(createEsrQrFromEsr9((XmlEsr9)body.getEsr(), qrIban))
+										.build())
+								.build())
+						.build());
+			}
+		}
+		return xAugmentedRequest;
+	}
+
+	private boolean isValidEsrQRCandidate(final XmlRequest xAugmentedRequest)
+	{
+		return hasEsr9(xAugmentedRequest) && isReferenceNumberConvertible(xAugmentedRequest);
+	}
+
+	/**
+	 * This check is needed because, based on specs, Esr9 reference numbers can be of either type:
+	 * <ol>
+	 *     <li>([0-9] [0-9]{5} [0-9]{5} [0-9]{5}</li>
+	 *     <li>[0-9]{2} [0-9]{5} [0-9]{5} [0-9]{5} [0-9]{5} [0-9]{5})</li>
+	 * </ol>
+	 * While the EsrQR reference number accepted patterns are:
+	 * <ol>
+	 * 	  <li>[0-9]{27}</li>
+	 * 	  <li>RF[0-9]{2}[0-9A-Z]{1,21}</li>
+	 * </ol>
+	 * As a result, the ESR9 reference number can be converted to EsrQR only if it's exactly 27 characters long.
+	 *
+	 * @param xAugmentedRequest the XmlRequest to check
+	 * @return true if the reference number is convertible to EsrQR, false otherwise.
+	 */
+	private boolean isReferenceNumberConvertible(@NonNull final XmlRequest xAugmentedRequest)
+	{
+		return StringUtils.cleanWhitespace(((XmlEsr9)xAugmentedRequest.getPayload()
+						.getBody()
+						.getEsr())
+						.getReferenceNumber())
+				.length() == 27;
+	}
+
+	private XmlEsr createEsrQrFromEsr9(final @NonNull XmlEsr9 esr9, @NonNull final String qrIban)
+	{
+		final XmlEsrQR.XmlEsrQRBuilder xEsrQR = XmlEsrQR.builder();
+
+		xEsrQR.bank(esr9.getBank());
+		xEsrQR.creditor(esr9.getCreditor());
+		xEsrQR.type(jaxbRequestObjectFactory.createEsrQRType().getType());
+		xEsrQR.iban(qrIban);
+		xEsrQR.referenceNumber(StringUtils.cleanWhitespace(esr9.getReferenceNumber()));
+		xEsrQR.paymentReason(Collections.emptyList());
+
+		return xEsrQR.build();
+	}
+
+	private boolean hasEsr9(final XmlRequest xAugmentedRequest)
+	{
+		return xAugmentedRequest.getPayload().getBody().getEsr() instanceof XmlEsr9;
+	}
+
+	@Nullable
+	private String getQrIbanOrNull(final BPartnerId bpartnerID)
+	{
+		final BPartnerBankAccount bPartnerBankAccount = bpBankAccountDAO.getBpartnerBankAccount(BankAccountQuery.builder()
+				.containsQRIBAN(true)
+				.bPartnerId(bpartnerID)
+				.build()).stream().findFirst().orElse(null);
```

Comment:
not aware about all functional specs, but wondering if we shall get the "first" bank account which has a QR IBAN set.

About "first"... can we have a predictable order?

Reply:
@teosarca afaict the default bank account shall be given priority, but other than that any QR IBAN will do.

Reply:
@adi-stefan (cc @teosarca ) don't give precedence just to default, but filter for the type (`BPBANKACCTUSE_DirectDeposit` and `BPBANKACCTUSE_Both`) 
Extent BankAccountQuery accordingly.
And make sure to order by ID, so it's predicatable

Reply:
Changed filter/sort criteria.